docs/NativeCodeGeneration.md

```diff
@@ -0,0 +1,185 @@
+equivalent C operations. In addition, the generated C file also contains bindings for any helper functions referenced
+by the eBPF program and any maps the eBPF program uses.
+
+## Step 4 – Compile generated C code into native COFF file
+
+The fourth step is to use a standard C compiler (GCC, MSVC, or Clang) to generate an x64/ARM64 COFF file with Windows
+calling conventions. The C file will expose a single top-level global variable that allows the skeleton code to find
+the helpers, maps, and programs contained in the file.
+
+## Step 5 – Linking the COFF file into a PE image
+
+The fifth is to use the platform-specific linker to generate the final PE file. The linker consumes the object file
+from step 4, a generic DriverEntry skeleton.
+
+## Step 6 – Signing the PE file
+
+The last step is to use the standard PE signing tools to produce a signed PE file from the output of step 5. For the
+developer workflow the signature would be based on a test signing root, whereas for production use the PE file would
+be signed using the Microsoft PRS signing keys.
+
+# Stub files consumed by the linker
+
+Step 5 consumes a set of object files to generate the final PE image. The following section gives a summary of the
+driver skeleton.
+
+## DriverEntry skeleton
+
+All Windows Drivers need to export a static function labelled &quot;DriverEntry&quot; that is invoked by the OS
+(Operating System) during loading of the driver. In addition to the standard driver related functionality, the driver
+entry skeleton would also register three [Network Module Registrar (NMR)](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/introduction-to-the-network-module-registrar)
+contracts exposed by the other framework components (to export the program to the eBPF runtime, to create any maps
+required, and to import the helper functions).
+
+## Global entry point for generated code
+
+Every generated C file will contain a single global entry point of type metadata_table_t:
+
+```
+typedef struct _metadata_table
+{
+    void (*programs)(program_entry_t**programs, size_t*count);
+    void (*maps)(map_entry_t**maps, size_t*count);
+    void (*helpers)(helper_function_entry_t**helpers, size_t*count);
+} metadata_table_t;
+
+metadata_table_t xdp = { _get_programs, _get_maps, _get_helpers };
+```
+
+The meta-data table provides pointers to three functions that permit querying the list of programs, the list of maps,
+and the list of helper functions. The table name is derived from the ELF file name, with _ replacing any character that
+is not valid in a C variable name. This variable is the only globally visible variable in the generated C code.
+
+## Exported programs
+
+Each program in the generated C file is exported via a program_entry_t:
+```
+typedef struct _program_entry
+{
+    uint64_t (*function)(void*);
+    const char* name;
+} program_entry_t;
```

Comment:
Just mentioning that this will require revision as this is already deviating from the code.